primitives/utility/src/lib.rs

```diff
@@ -51,3 +61,191 @@ impl<T: UpwardMessageSender, W: WrapVersion> SendXcm for ParentAsUmp<T, W> {
 		}
 	}
 }
+
+/// Contains information to handle refund/payment for xcm-execution
+#[derive(Clone, Eq, PartialEq, Debug)]
+struct AssetTraderRefunder<Balance, AssetId> {
+	// The amount of weight bought minus the weigh already refunded
+	weight: Weight,
+	// The balance paid for the weight bought minus the balance already refunded
+	balance: Balance,
+	// The multilocation representing the fungible asset in which fees were paid
+	location: MultiLocation,
+	// The local assetId the aforementioned location corresponds to
+	asset: AssetId,
+}
+
+/// Charges for exercution in the first multiasset of those selected for fee payment
+/// Only succeeds for Concrete Fungible Assets
+/// First tries to convert the this MultiAsset into a local assetId
+/// Then charges for this assetId as described by FeeCharger
+/// Weight, paid balance, local asset Id and the multilocation is stored for
+/// later refund purposes
+/// Important: Errors if the Trader is being called twice by 2 BuyExecution instructions
+/// Alternatively we could just return payment in the aforementioned case
+pub struct TakeFirstAssetTrader<
+	AccountId,
+	FeeCharger: ChargeWeightInFungibles<AccountId, Assets>,
+	Matcher: MatchesFungibles<Assets::AssetId, Assets::Balance>,
+	Assets: fungibles::Mutate<AccountId> + fungibles::Transfer<AccountId> + fungibles::Balanced<AccountId>,
```

Comment:
```suggestion
	ConcreteAssets: fungibles::Mutate<AccountId> + fungibles::Transfer<AccountId> + fungibles::Balanced<AccountId>,
```

to better communicate intent.